### ISSUE TITLE
Add: Plot - Spectra Plot

### ISSUE THREAD

**Issue by dyrlund** [mzmine/mzmine3#12]

**Comment by dyrlund** [comment]
Initial version added in #74 and #75.

**Comment by dyrlund** [comment]
@tomas-pluskal On my system, the JAVAFX_CHARTS is very slow to load data into the chart - it takes about 14 seconds to load one MS spectrum - and I'm experiencing lag when selecting an area to zoom. Do you have the same issue?

**Comment by tomas-pluskal** [comment]
I guess it depends on how many data points you have in the spectrum. 
JavaFX charts require creation of many objects (each data point is an individual Node). 
On the contrary, JFreeChart is backed by primitive double[] arrays so it is very efficient. 
Perhaps it might be better to stick to JFreeChart for scientific plots. Unfortunately, the JavaFX version of JFreeChart lacks some features of the Swing version :(

**Comment by tomas-pluskal** [comment]
Thomas, what is your impression from the 3 versions of the spectrum plot?

**Comment by dyrlund** [comment]
From a design point of view I like the JAVAFX_CHARTS the most. It fits nicely with the new JavaFX interface we are using. I also think it is an advantage that it is a new library which will be developed further. It will however require that we can fix the loading issue.

As for the two others, then I prefer JFREECHART over WATERLOOFX. JFREECHART will give a very similar result to the plots from MZmine 2.

There are some issues with them all which we will have to fix depending on which we select, here is what I have found so far.

**JAVAFX_CHARTS**
- Slow to load data
- Y-axis numbers are not formatted
- Too many x-axis numbers when you zoom

**JFREECHART**
- Peaks are connected instead of being represented as a single line

**WATERLOOFX**
- Peaks are connected instead of being represented as a single line
- Zooming is weird
- Axis numbers are not formatted


Have you looked into how easy it is to overlay multiple plots in the same figure with the different libraries? I think this would be a nice extra functionality in MZmine 3.

**Comment by dyrlund** [comment]
Formatting the y-axis in the JAVAFX_CHARTS can be achieved using the following code:

	NumberFormat intensityFormat = MZmineCore.getConfiguration().getIntensityFormat();
	yAxis.setTickLabelFormatter(new StringConverter<Number>() {
		@Override
		public String toString(Number number) {
			return intensityFormat.format(number.doubleValue());
		}

		@Override
		public Number fromString(String string) {
			try {
				return intensityFormat.parse(string);
			} catch (ParseException e) {
				e.printStackTrace();
				return 0;
			}
		}
	});

**Comment by dyrlund** [comment]
I tried using an ObservableList for the JAVAFX_CHARTS and it seems to speed up the loading of the data a bit, but it is still slow.

I did however debug the slow speed to the tooltip code. Removing the following code makes the plot speed equal to the JFREECHART:

	Text labelNode = new Text();
	labelNode.getStyleClass().add("chart-item-label");
	String tooltipText = "m/z: "
			+ MZmineCore.getConfiguration().getMZFormat().format(mz)
			+ "\nIntensity: " + MZmineCore.getConfiguration()
					.getIntensityFormat().format(intensity);
	Tooltip.install(labelNode, new Tooltip(tooltipText));
	newData.setNode(labelNode);

**Comment by tomas-pluskal** [comment]
It is possible to add multiple data series to both JavaFX charts and JFreeChart. But JFreeChart is much more flexible in terms of rendering the series (e.g. rendering centroid and profile data in a single plot), changing colors etc. JFreeChart is also way more memory efficient. To be honest, after playing with the JavaFX charts API for a while, I feel a bit bitter about it.

Of course, we can use JavaFX charts for data analysis plots, like scatter plot or cloud plot, but I think it is not a good library for rendering raw data.

**Comment by dyrlund** [comment]
Lets go with the JFreeChart for the charts which require rendering large amounts of raw data and then we can always upgrade to the JavaFX chart when this library has improved more.

What about the 3D plots?

**Comment by tomas-pluskal** [comment]
I have several options in mind for the 3D plot, need to test them a bit and see.

**Comment by dyrlund** [comment]
With regards to the icons - what do you think about using, when possible, the [Font Awesome Icons](https://fortawesome.github.io/Font-Awesome/) in MZmine 3?

**Comment by tomas-pluskal** [comment]
Sure, there is also this nice collection of icons with JavaFX and FXML support (includes Font Awesome, too): http://aalmiray.github.io/ikonli/

Still, I think we will need custom icons for things like mass spectra.

**Comment by dyrlund** [comment]
One feature which I think will be very useful in the spectrum plot, is an option to display data in a relative intensity instead of the absolute value. This will make it easier to compare data when multiple spectra are shown in the same plot. How easy do you think it would be to implement this?

**Comment by tomas-pluskal** [comment]
I am thinking to have a setup dialog where you can configure all layers shown in the spectrum plot. In the dialog, you would be able to set 
* color for each spectrum (layer)
* whether the spectrum is rendered as profile or centroided
* normalization factor for intensity (disabled by default, but can be normalized to any value)

If you have any other idea what would be good to include in the spectrum plot, now is the time :)

By the way, did you check the MsSpectrumPlotWindow.fxml file? I think it is pretty cool that we can define the whole layout in a single FXML file, including context menus, icons, etc.

**Comment by dyrlund** [comment]
I think it would also be nice to be able to easily change between the plot type ("base peak intensity" or "total ion current"). This could be as a button or through the setup dialog. 

It's amazing how much simpler it is with JavaFX - a great improvement!

**Comment by tomas-pluskal** [comment]
For the "base peak intensity" or "total ion current" switching, I suppose you are referring to the TIC (chromatogram) plot. Or does it make sense to have such selection for the spectrum plot?

**Comment by dyrlund** [comment]
Yes, sorry, I was specifically thinking of the TIC plots. I agree that I can't see a use for it under the spectrum plot.

**Comment by dyrlund** [comment]
I really like the new spectra plot which you have made. It is a great improvement over MZmine 2. Here is what I have found so far.

**Bugs:**

1. If you normalize the intensity scale and move to a new spectrum using the errors, then the intensity scale is reset.

2. If you overlay two or more spectra then they are all changed when using the errors. Maybe only the first or last spectrum should be changed.

3. When exporting an image, the "Save as type" is empty in the save as window.

4. If you zoom far enough in then you can get to negative intensity values. Is it possible to set the low limit of the y-axis to never go below 0?

5. The quality of the prints are very bad compared to MZMine 2. This is both the case for a real print and for a print using Adobe PDF. Also, the option to setup the page before printing (size and orientation) is missing.

**Suggestions:**

1. Add right click menu option to delete a layer. This currently has to be done through the setup layers menu.

2. When normalizing the intensity, it would be nice if the intensity scale was shown as 100 instead of 1.00E2.

3. In MZmine 2, clicking on the y-axis would set the intensity scale to show all the data in the current zoom range. Right now you reset the zoom by clicking on the y-axis.

4. You can add the same scan multiple times - does this make sense?

5. We probably need a button to hide the legend if many spectra are shown in the same plot. I also suggest that the legend is auto hid when more than e.g. 30 items are shown.

6. I think we should update the default colours we use to be colour blind friendly. The following eight colors is one example and are taken from http://jfly.iam.u-tokyo.ac.jp/color/:

- #999999 - Gray
- #E69F00 - Orange
- #56B4E9 - Sky Blue
- #009E73 - Bluish green
- #F0E442 - Yellow
- #0072B2 - Blue
- #D55E00 - Vermilion
- #CC79A7 - Reddish purple

**Comment by tomas-pluskal** [comment]
> If you overlay two or more spectra then they are all changed when using the errors. Maybe only the first or last spectrum should be changed.

That's a feature, not a bug!

>  When normalizing the intensity, it would be nice if the intensity scale was shown as 100 instead of 1.00E2.

The formatting is set globally in the preferences. We could probably add a more intelligent logic to the NumberFormat in the preferences (e.g., if the value is <1000.0, do not show the exponent), but I think it is not a problem of the spectrum plot.

>  In MZmine 2, clicking on the y-axis would set the intensity scale to show all the data in the current zoom range. Right now you reset the zoom by clicking on the y-axis.

Is there such function in MZmine 2, really? Didn't you make it up? :)
I think resetting the zoom of each axis by clicking on it is a good idea. I suggested this in JFreeChart forum, so it might become a default behavior of JFreeChart.

> You can add the same scan multiple times - does this make sense?

Probably not much, but it doesn't hurt either.

> I think we should update the default colours we use to be colour blind friendly.

I think I'll wait for some color-blind user to submit a pull request... :)

**Comment by dyrlund** [comment]
I agree that it would be a nice improvement to the NumberFormat if values below 1000 don't show the exponent. We could also consider allowing the user to set the specific value. Will you make the change?

And with regards to the zoom levels then I added this feature for the chromatogram plots in MZmine 2. I actually thought that I had added it to all the plots but apparently not. But I really like the feature so lets hope that it will be added to JFreeChart.

**Comment by dyrlund** [comment]
A few additional issues and comments:

1. The m/z axis always starts from 0 regardless of if there is only data from e.g. 100 or 200 m/z.

2. Why not allow to specify the width of the line for centroided data like it is possible to profile data?

3. When copying the spectra or exporting the spectra, it would be nice if it was possible to set a noise cut-off value so that everything below this is not included. As an example, I have a spectra with 4554 data points. If I limit this to only those with min. 1% of the max. intensity then I have 281 data points which is more relevant when using the spectra in other software. Another option would be to e.g. allow to select to only include the top 100 data points or something similar.

**Comment by tomas-pluskal** [comment]
Ad 1: I don't see that - in my case the X axis starts with the lowest m/z. Can you check if this might be specific to some raw files?

Ad 2: It's a different Renderer. I have not yet found a way how to do it. OK I'll search a bit more.

Ad 3. The thing is there are too many options how to do this. In my view, it is simple enough to export all data and the user can then filter them afterwards (top 100, top 1%, above certain intensity, etc.)

**Comment by dyrlund** [comment]
I assumed that it was always showing 0 but looking closer, I can see that I was wrong. It is not showing 0 but also not the lowest m/z value.

Here is an example for one of our mzML files where the lowest data point is 100 m/z:
![min 100 mz](https://cloud.githubusercontent.com/assets/10582167/13498176/371586ba-e159-11e5-944a-7985684bc4f5.png)

And here is the orbitrap_300-600mz test file where the lowest data point is 300 m/z:
![min 300 mz](https://cloud.githubusercontent.com/assets/10582167/13498177/3807f9ea-e159-11e5-8576-4554c35fb32a.png)

**Comment by tomas-pluskal** [comment]
Yes, it is adding some % margin to the lowest and highest values, because otherwise the label for the lowest value would often be half-visible (see the 300.0000 label in your second plot - if we start the plot exactly at 300.0000, half of the label would disappear). It's a cosmetic thing.

**Comment by dyrlund** [comment]
I see. It just looks weird that there is a gap in the beginning. And there is no way to offset the labels based on the y-axis position? Otherwise that could be a nice improvement to JFreeChart.

**Comment by tomas-pluskal** [comment]
JFreeChart is now on github, so you can submit a pull request there :)
